src/cls/rgw/cls_rgw.cc

```diff
@@ -1172,18 +1172,32 @@ int rgw_bucket_complete_op(cls_method_context_t hctx, bufferlist *in, bufferlist
   if (op.op == CLS_RGW_OP_CANCEL) {
     log_op = false; // don't log cancelation
     if (op.tag.size()) {
-      // we removed this tag from pending_map so need to write the changes
-      CLS_LOG_BITX(bitx_inst, 20,
-		   "INFO: %s: setting map entry at key=%s",
-		   __func__, escape_str(idx).c_str());
-      bufferlist new_key_bl;
-      encode(entry, new_key_bl);
-      rc = cls_cxx_map_set_val(hctx, idx, &new_key_bl);
-      if (rc < 0) {
-	CLS_LOG_BITX(bitx_inst, 1,
-		     "ERROR: %s: unable to set map val, key=%s, rc=%d",
-		     __func__, escape_str(idx).c_str(), rc);
-        return rc;
+      if (!entry.exists && entry.pending_map.empty()) {
+        // a racing delete succeeded, and we canceled the last pending op
```

Comment:
I don't believe we can know that it was necessarily a racing delete here since CLS_RGW_OP_CANCEL is generic over the type of op being canceled. i.e. it looks like it's possible for the same thing to occur if we were to have an CLS_RGW_OP_ADD that gets cancelled. I think this new logic is sound in either case, but may want to update the comment.

Reply:
`!entry.exists` implies that the last successful complete_op was a delete. the comment doesn't necessarily say that *we're* a delete

Reply:
`rgw_bucket_prepare_op` sets entry.exists to false and it isn't set to true until `rgw_bucket_complete_op` successfully completes the CLS_RGW_OP_ADD, so I don't think that `!entry.exists` implies that there was ever a previous successful complete_op, right?

Reply:
you're right; but with `pending_map.empty()` as well, i think it's correct?

Reply:
I think the pending_map is cleared in either case here: https://github.com/ceph/ceph/pull/50041/files#diff-00c1ab51461aa6cc822120d41afc7cc805e1cef7c8cc185d40147d8ffa1bea79L1154 ?

Reply:
Isn't just one entry removed from the pending map on line 1154? And that doesn't necessarily mean (although it could mean) the pending map is now empty.